### DISCUSSION TITLE
Interface with default fun implementations as a easier way to add custom conversions?

### DISCUSSION THREAD

**Discussion by jakoss** [mcarleio/konvert#10] [Ideas]
Current way of adding custom type converters requires a separate module, using `AutoService` and all that. Library called `kmapper` is using a simpler approach. It's using default interface fun implementations to specify converters. It's super easy that way. Maybe you can consider doing something similar?

Example: mapper using converter: https://github.com/s0nicyouth/kmapper/blob/master/examples/testload/src/main/kotlin/com/syouth/kmapper/testload/mappers/interfaceConverterTest/InterfaceConverterMapper.kt and the converter implementation: https://github.com/s0nicyouth/kmapper/blob/master/converters/src/main/kotlin/com/syouth/kmapper/converters/PrimitiveTypeConverters.kt

**Comment by jakoss** [comment]
Possible usecase for me is the following: convert `OffsetDateTime` to `LocalDateTime`. Interface would look like this:

```kotlin
import java.time.*

/**
 * In our case the [OffsetDateTime] we're getting from the api represents time in UTC.
 * We can convert that to [LocalDateTime] just by moving that by the offset of current zone to UTC ([localZoneOffset] field stores that offset).
 */
interface DateTimeConverters {
    fun asOffsetDateTime(localDateTime: LocalDateTime?): OffsetDateTime? =
        localDateTime?.atOffset(localZoneOffset)

    companion object {
        val localZoneOffset: ZoneOffset by lazy { OffsetDateTime.now().offset }
    }
}
```

But to implement that as converter in the Konvert is much more tricky, since the `localZoneOffset` cannot be shared easily (and that's pretty significant performance optimization)

**Reply by mcarleio** [reply]
I think this is all already possible:

1. Providing your own function in any `@Konverter` will register a new type converter which has higher priority than any default existing type converters.

   ```kotlin
   @Konverter
   interface Mapper {
       @Konvert
       fun toTarget(source: SourceDate): TargetDate
   
       fun asOffsetDateTime(source: LocalDateTime): OffsetDateTime = source.atOffset(localZoneOffset)
   
       companion object {
           val localZoneOffset: ZoneOffset by lazy { OffsetDateTime.now().offset }
       }
   }
   
   class SourceDate(val dateTime: LocalDateTime)
   class TargetDate(val dateTime: OffsetDateTime)
   ```
   (!) This type converter will be used for all mappings from `LocalDateTime` to `OffsetDateTime`.

2. Implement the mapping individually only for the one mapping (like you described already):
   ```kotlin
   @Konverter
   interface Mapper: MyTemporalMapper {
       @Konvert(mappings = [
           Mapping(target = "dateTime", expression = "it.dateTime.asOffsetDateTime()")
       ])
       fun toTarget(source: SourceDate): TargetDate
   }
   
   interface MyTemporalMapper {
       fun LocalDateTime.asOffsetDateTime(): OffsetDateTime = this.atOffset(localZoneOffset)
   
       companion object {
           val localZoneOffset: ZoneOffset by lazy { OffsetDateTime.now().offset }
       }
   }
   
   class SourceDate(val dateTime: LocalDateTime)
   class TargetDate(val dateTime: OffsetDateTime)
   ```
   Instead of `interface MyTemporalMapper` you could also write extension functions, objects, ... whatever. In an expression you can do whatever you want.


Reminder for myself: I have to document all this stuff somewhere to show how easy it is :smile:

**Reply by jakoss** [reply]
So it's almost perfect, just the part about manually mapping it only for one mapping. My use case is to inherit from that interface and then don't have to bother about specifying it. So just inheriting would be like having converter defined in the SPI

**Reply by jakoss** [reply]
Or maybe i misunderstood - if i declare it like that:

```kotlin
@Konverter
interface DateTimeConverters {
    fun asOffsetDateTime(source: LocalDateTime): OffsetDateTime = source.atOffset(localZoneOffset)

    companion object {
        val localZoneOffset: ZoneOffset by lazy { OffsetDateTime.now().offset }
    }
}
```

that will automatically be used by other mappers? Even in another modules?

**Reply by mcarleio** [reply]
Ok, so you mean that if your `@Konverter` interface extends some interface like `MyTemporalMapper`, then it should register the functions from that only for the scope of this `@Konverter`, right?

**Reply by mcarleio** [reply]
> Or maybe i misunderstood - if i declare it like that:
> 
> ```kotlin
> @Konverter
> interface DateTimeConverters {
>     fun asOffsetDateTime(source: LocalDateTime): OffsetDateTime = source.atOffset(localZoneOffset)
> 
>     companion object {
>         val localZoneOffset: ZoneOffset by lazy { OffsetDateTime.now().offset }
>     }
> }
> ```
> 
> that will automatically be used by other mappers? Even in another modules?

Yes, this will register a `TypeConverter` for `LocalDateTime` to `OffsetDateTime` which is then used for all mappings from those types.

**Reply by jakoss** [reply]
> > Or maybe i misunderstood - if i declare it like that:
> > ```kotlin
> > @Konverter
> > interface DateTimeConverters {
> >     fun asOffsetDateTime(source: LocalDateTime): OffsetDateTime = source.atOffset(localZoneOffset)
> > 
> >     companion object {
> >         val localZoneOffset: ZoneOffset by lazy { OffsetDateTime.now().offset }
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > that will automatically be used by other mappers? Even in another modules?
> 
> Yes, this will register a `TypeConverter` for `LocalDateTime` to `OffsetDateTime` which is then used for all mappings from those types.

Ok, it works, but i still have to inherit from that interface. Without that Konvert cannot find the converter itself. But it's perfect as it is - good thing to document :)

**Reply by jakoss** [reply]
Oh no, it's not perfect. It compiled somewhat strange. So i have that for a test:

```kotlin
// Core module
@Konverter
interface DateTimeConverters {
    fun asOffsetDateTime(source: LocalDateTime): OffsetDateTime = source.atOffset(localZoneOffset)

    companion object {
        val localZoneOffset: ZoneOffset by lazy { OffsetDateTime.now().offset }
    }
}

// weather module
@Konverter
interface WeatherMapper : DateTimeConverters {
    fun fromDto(weatherDto: CurrentWeatherDto): CurrentWeather
    fun toDto(weather: CurrentWeather): CurrentWeatherDto
}
```

Generated code looks like this:
```kotlin
public class WeatherMapperImpl : WeatherMapper {
  public override fun fromDto(weatherDto: CurrentWeatherDto): CurrentWeather = CurrentWeather(
    temperature = weatherDto.temperature,
    windspeed = weatherDto.windspeed,
    time = weatherDto.time.toLocalDateTime()
  )

  public override fun toDto(weather: CurrentWeather): CurrentWeatherDto = CurrentWeatherDto(
    temperature = weather.temperature,
    windspeed = weather.windspeed,
    time = io.mcarle.konvert.api.Konverter.get<pl.test.data.weather.impl.WeatherMapper>().asOffsetDateTime(source = weather.time)
  )
}
```

And when i try to call `toDto` i have exception:
```
java.util.NoSuchElementException: No element of the collection was transformed to a non-null value.
    at io.mcarle.konvert.api.Konverter$Companion.get(Konverter.kt:51)
    at pl.test.data.weather.impl.WeatherMapperImpl.toDto(WeatherMapperKonverter.kt:19)
```

@mcarleio am i doing something wrong is should i create an issue for that?

If i'm not inheriting from the interface then i have ksp error:
```
e: [ksp] java.util.NoSuchElementException: Could not find converter for weather.time -> time: LocalDateTime -> OffsetDateTime
	at io.mcarle.konvert.processor.codegen.MappingCodeGenerator$convertValue$1.invoke(MappingCodeGenerator.kt:186)
	at io.mcarle.konvert.processor.codegen.MappingCodeGenerator$convertValue$1.invoke(MappingCodeGenerator.kt:183)
	at io.mcarle.konvert.converter.api.TypeConverterRegistry.withAdditionallyEnabledConverters(TypeConverterRegistry.kt:47)
	at io.mcarle.konvert.processor.codegen.MappingCodeGenerator.convertValue(MappingCodeGenerator.kt:183)
	at io.mcarle.konvert.processor.codegen.MappingCodeGenerator.constructorParamsCode(MappingCodeGenerator.kt:56)
	at io.mcarle.konvert.processor.codegen.MappingCodeGenerator.constructorCode(MappingCodeGenerator.kt:37)
	at io.mcarle.konvert.processor.codegen.MappingCodeGenerator.generateMappingCode(MappingCodeGenerator.kt:24)
	at io.mcarle.konvert.processor.codegen.CodeGenerator.generateCode(CodeGenerator.kt:43)
	at io.mcarle.konvert.processor.konvert.KonverterCodeGenerator$generate$3$1.invoke(KonverterCodeGenerator.kt:74)
	at io.mcarle.konvert.processor.konvert.KonverterCodeGenerator$generate$3$1.invoke(KonverterCodeGenerator.kt:45)
	at io.mcarle.konvert.converter.api.config.ConfigurationKt.withIsolatedConfiguration(Configuration.kt:47)
	at io.mcarle.konvert.processor.konvert.KonverterCodeGenerator.generate(KonverterCodeGenerator.kt:45)
	at io.mcarle.konvert.processor.KonvertProcessor$generateMappingCode$1$1.invoke(KonvertProcessor.kt:64)
	at io.mcarle.konvert.processor.KonvertProcessor$generateMappingCode$1$1.invoke(KonvertProcessor.kt:60)
	at io.mcarle.konvert.converter.api.config.ConfigurationKt.withIsolatedConfiguration(Configuration.kt:47)
	at io.mcarle.konvert.processor.KonvertProcessor.generateMappingCode(KonvertProcessor.kt:60)
	at io.mcarle.konvert.processor.KonvertProcessor.access$generateMappingCode(KonvertProcessor.kt:24)
	at io.mcarle.konvert.processor.KonvertProcessor$process$1.invoke(KonvertProcessor.kt:40)
	at io.mcarle.konvert.processor.KonvertProcessor$process$1.invoke(KonvertProcessor.kt:33)
	at io.mcarle.konvert.converter.api.config.ConfigurationKt.withIsolatedConfiguration(Configuration.kt:41)
	at io.mcarle.konvert.processor.KonvertProcessor.process(KonvertProcessor.kt:33)
```

**Reply by mcarleio** [reply]
Ok, so regarding why you have to inherit at the moment: This has to do with what I wrote here: https://github.com/mcarleio/konvert/discussions/11#discussioncomment-5839409

Regarding that java.util.NoSuchElementException, please create an issue, I am unsure about that.
